### PR TITLE
[FIX] bus: reconnect websocket after user log in/out

### PR DESCRIPTION
**Pull request by tsm-odoo**
Before this PR, a websocket would wait for the server to close
it with a `SessionExpired` close code to be refreshed after the user
logged in/out. Indeed, the server is responsible to check for outdated
sessions on incoming/outgoing messages.

This is problematic: a user logging in with more than one tab opened,
would have to wait to receive its messages. This PR solves this issue
by refreshing the connection immediately in this scenario.

### CODE REVIEW

File: addons/bus/static/src/workers/websocket_worker.js [hunk 1]
```diff
@@ -193,11 +194,22 @@ export class WebsocketWorker {
      * given client.
      * @param {Number} [param0.lastNotificationId] Last notification id
      * known by the client.
+     * @param {Number|undefined} [param0.uid] Current user id.
```
Comment (seb-odoo) [inline addons/bus/static/src/workers/websocket_worker.js:197]: It can be `undefined`?

File: addons/bus/static/src/workers/websocket_worker.js [hunk 2]
```diff
@@ -193,11 +194,22 @@ export class WebsocketWorker {
...
      */
-    _initializeConnection(client, { debug, lastNotificationId }) {
+    _initializeConnection(client, { debug, lastNotificationId, uid }) {
         this.lastNotificationId = lastNotificationId;
         this.debugModeByClient[client] = debug;
         this.isDebug = Object.values(this.debugModeByClient).some(debugValue => debugValue !== '');
+        if (!this._isWebsocketConnected()) {
+            return;
+        }
```
Comment (seb-odoo) [inline addons/bus/static/src/workers/websocket_worker.js:205]: Who is this related to change with user?

File: addons/bus/static/src/workers/websocket_worker.js (hunk 1 above)
Comment (tsm-odoo) [reply addons/bus/static/src/workers/websocket_worker.js:197]: If no user is connected, yes !

File: addons/bus/static/src/workers/websocket_worker.js (hunk 2 above)
Comment (tsm-odoo) [reply addons/bus/static/src/workers/websocket_worker.js]: I slightly changed the syntax but the point is not closing a socket which is not yet connected (it would lead to warnings).

File: addons/bus/static/tests/bus_tests.js [hunk 3]
```diff
@@ -423,6 +423,66 @@ QUnit.module('Bus', {
...
+        patchWithCleanup(session, {
+            uid: undefined,
+        });
+        await makeTestEnv();
+        await disconnectionDoneDeferred;
+
+        assert.verifySteps([`disconnect - ${WEBSOCKET_CLOSE_CODES.SESSION_EXPIRED}`]);
+    });
+
+    QUnit.test('Websocket connection is refreshed upon user log out', async function (assert) {
```
Comment (seb-odoo) [inline addons/bus/static/tests/bus_tests.js:456]: name of tests are inverted to what is done inside

Comment (tsm-odoo) [reply addons/bus/static/tests/bus_tests.js]: updated :+1:

**Review by seb-odoo — APPROVED** [review]
@robodoo r+

It would be nice to prevent obsolete tabs from receiving the new notifications too but current PR is already a good improvement.

File: addons/bus/static/src/services/bus_service.js [hunk 4]
```diff
@@ -29,6 +30,7 @@ export const busService = {
         const worker = new workerClass('/bus/websocket_worker_bundle', {
             name: 'SharedWorker' in window ? 'odoo:websocket_shared_worker' : 'odoo:websocket_worker',
         });
+        const currentUID = (Array.isArray(session.user_id) ? session.user_id[0] : session.user_id);
```
Comment (seb-odoo) [inline addons/bus/static/src/services/bus_service.js:33]: Why 2 formats?

File: addons/bus/static/src/services/bus_service.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
 /** @odoo-module **/
 
 import { browser } from "@web/core/browser/browser";
+import { SessionExpiredDialog } from "@web/core/errors/error_dialogs";
```
Comment (seb-odoo) [inline addons/bus/static/src/services/bus_service.js:4]: Is that available in all bundles were bus service is used?

File: addons/bus/static/src/services/bus_service.js [hunk 6]
```diff
@@ -60,6 +62,11 @@ export const busService = {
             if (type === 'notification') {
                 multiTab.setSharedValue('last_notification_id', data[data.length - 1].id);
                 data = data.map(notification => notification.message);
+            } else if (type === 'session_expired' && (data.uid === currentUID || currentUID === undefined)) {
```
Comment (seb-odoo) [inline addons/bus/static/src/services/bus_service.js:65]: What is the `undefined` case?

File: addons/bus/static/src/workers/websocket_worker.js [hunk 7]
```diff
@@ -193,12 +196,23 @@ export class WebsocketWorker {
      * given client.
      * @param {Number} [param0.lastNotificationId] Last notification id
      * known by the client.
+     * @param {Number|false|undefined} [param0.uid] Current user id.
```
Comment (seb-odoo) [inline addons/bus/static/src/workers/websocket_worker.js:199]: This is getting really messy, it would be nice to explain each type.

File: addons/bus/static/src/workers/websocket_worker.js [hunk 8]
```diff
@@ -244,6 +258,10 @@ export class WebsocketWorker {
             // Don't wait to reconnect on keep alive timeout.
             this.connectRetryDelay = 0;
         }
+        if (code === WEBSOCKET_CLOSE_CODES.SESSION_EXPIRED) {
+            this.isWaitingForNewUID = !reason || !JSON.parse(reason).closedByClient;
+            this.broadcast('session_expired', { uid: this.isWaitingForNewUID ? this.currentUID : this.lastUID });
```
Comment (seb-odoo) [inline addons/bus/static/src/workers/websocket_worker.js:263]: The difference between when to use current or last is not very clear.

File: addons/bus/static/src/services/bus_service.js (hunk 4 above)
Comment (tsm-odoo) [reply addons/bus/static/src/services/bus_service.js:32]: In the frontend session info, it's a number while in the backend an array containing only one number..

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
